### PR TITLE
[FAW] parallel FreqAwareEmbedding

### PR DESCRIPTION
**Pull request by feifeibear**

### CODE REVIEW

**Comment by feifeibear** [comment]
Copy from https://github.com/zxgx/FreqCacheEmbedding/blob/main/recsys/modules/embeddings/freq_aware_embedding.py

**Review by zxgx — APPROVED** [review]

File: colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    def forward(self, indices, offsets=None, per_sample_weights=None, shape_hook=None, scatter_dim=0, gather_dim=-1):
+        with torch.no_grad():
+            reorder_ids = self.cache_weight_mgr.prepare_ids(indices)
+
+        output_shard = F.embedding_bag(reorder_ids, self.cache_weight_mgr.cuda_cached_weight, offsets, self.max_norm,
+                                       self.norm_type, self.scale_grad_by_freq, self.mode, self.sparse,
+                                       per_sample_weights, self.include_last_offset, self.padding_idx)
+
+        if shape_hook is not None:
+            output_shard = shape_hook(output_shard)
```
Comment (zxgx) [inline colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:108]: the shape hook might introduce some tensor view&squeeze operations.
I suppose we should be aware of that when using ColoParameter.

Comment (feifeibear) [reply colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:108]: OK, can you update it in another PR? I copy this line from yours code.

Comment (zxgx) [reply colossalai/nn/_ops/cache_embedding/parallel_freq_aware_embedding.py:108]: OK

**Pull request merged**